libfwupd/fwupd-device.c

```diff
@@ -1612,6 +1612,99 @@ fwupd_pad_kv_ups (GString *str, const gchar *key, FwupdUpdateState value)
+		json_builder_set_member_name (builder, FWUPD_RESULT_KEY_GUID);
+		json_builder_begin_array (builder);
+		json_builder_begin_object (builder);
+		for (guint i = 0; i < priv->guids->len; i++) {
+			const gchar *guid = g_ptr_array_index (priv->guids, i);
+			g_autofree gchar *title = g_strdup_printf ("%s%u", FWUPD_RESULT_KEY_GUID, i);
+			fwupd_device_json_add_string (builder, title, guid);
+		}
+		json_builder_end_object (builder);
+		json_builder_end_array (builder);
+	}
+	fwupd_device_json_add_string (builder, FWUPD_RESULT_KEY_SERIAL, priv->serial);
+	fwupd_device_json_add_string (builder, FWUPD_RESULT_KEY_SUMMARY, priv->summary);
+	fwupd_device_json_add_string (builder, FWUPD_RESULT_KEY_DESCRIPTION, priv->description);
+	fwupd_device_json_add_string (builder, FWUPD_RESULT_KEY_PLUGIN, priv->plugin);
+	fwupd_device_json_add_int (builder, FWUPD_RESULT_KEY_FLAGS, priv->flags);
+	if (priv->checksums->len > 0) {
+		json_builder_set_member_name (builder, "Checksums");
+		json_builder_begin_array (builder);
+		json_builder_begin_object (builder);
+		for (guint i = 0; i < priv->checksums->len; i++) {
+			const gchar *checksum = g_ptr_array_index (priv->checksums, i);
+			g_autofree gchar *checksum_display = fwupd_checksum_format_for_display (checksum);
+			g_autofree gchar *title = g_strdup_printf ("%s%u", FWUPD_RESULT_KEY_CHECKSUM, i);
+			fwupd_device_json_add_string (builder, title, checksum_display);
+		}
+		json_builder_end_object (builder);
+		json_builder_end_array (builder);
+	}
+	fwupd_device_json_add_string (builder, FWUPD_RESULT_KEY_VENDOR, priv->vendor);
+	fwupd_device_json_add_string (builder, FWUPD_RESULT_KEY_VENDOR_ID, priv->vendor_id);
+	fwupd_device_json_add_string (builder, FWUPD_RESULT_KEY_VERSION, priv->version);
+	fwupd_device_json_add_string (builder, FWUPD_RESULT_KEY_VERSION_LOWEST, priv->version_lowest);
+	fwupd_device_json_add_string (builder, FWUPD_RESULT_KEY_VERSION_BOOTLOADER, priv->version_bootloader);
+	fwupd_device_json_add_int (builder, FWUPD_RESULT_KEY_FLASHES_LEFT, priv->flashes_left);
+	if (priv->icons->len > 0) {
+		g_autoptr(GString) tmp = g_string_new (NULL);
+		for (guint i = 0; i < priv->icons->len; i++) {
+			const gchar *icon = g_ptr_array_index (priv->icons, i);
+			g_string_append_printf (tmp, "%s,", icon);
+		}
+		if (tmp->len > 1)
+			g_string_truncate (tmp, tmp->len - 1);
+		fwupd_device_json_add_string (builder, FWUPD_RESULT_KEY_ICON, tmp->str);
+	}
+	fwupd_device_json_add_int (builder, FWUPD_RESULT_KEY_INSTALL_DURATION, priv->install_duration);
+	fwupd_device_json_add_int (builder, FWUPD_RESULT_KEY_CREATED, priv->created);
+	fwupd_device_json_add_int (builder, FWUPD_RESULT_KEY_MODIFIED, priv->modified);
+	fwupd_device_json_add_int (builder, FWUPD_RESULT_KEY_UPDATE_STATE, priv->update_state);
+	fwupd_device_json_add_string (builder, FWUPD_RESULT_KEY_UPDATE_ERROR, priv->update_error);
+	fwupd_device_json_add_string (builder, FWUPD_RESULT_KEY_UPDATE_MESSAGE, priv->update_message);
+}
```

Comment:
Don't we also need `fwupd_release_to_json()` being called here (like `fwupd_device_to_string()` does)? I agree for the case you're using it it doesn't make much sense, but for adding general purpose API like this it seems sensible.

Reply:
Sure, added it.